apps/prairielearn/src/pages/publicAssessments/publicAssessments.html.ts

```diff
@@ -0,0 +1,93 @@
+import { z } from 'zod';
+
+import { html } from '@prairielearn/html';
+
+import { AssessmentModuleHeading } from '../../components/AssessmentModuleHeading.html.js';
+import { AssessmentSetHeading } from '../../components/AssessmentSetHeading.html.js';
+import { PageLayout } from '../../components/PageLayout.html.js';
+import {
+  AssessmentModuleSchema,
+  AssessmentSchema,
+  AssessmentSetSchema,
+} from '../../lib/db-types.js';
+
+export const AssessmentRowSchema = AssessmentSchema.extend({
+  start_new_assessment_group: z.boolean(),
+  assessment_set: AssessmentSetSchema,
+  assessment_module: AssessmentModuleSchema,
+  label: z.string(),
+});
+type AssessmentRow = z.infer<typeof AssessmentRowSchema>;
+
+export function PublicAssessments({
+  resLocals,
+  rows,
+  assessmentsGroupBy,
+}: {
+  resLocals: Record<string, any>;
+  rows: AssessmentRow[];
+  assessmentsGroupBy: 'Set' | 'Module';
+}) {
+  return PageLayout({
+    resLocals,
+    pageTitle: 'Assessments',
+    navContext: {
+      type: 'public',
+      page: 'assessments',
+    },
+    options: {
+      fullWidth: false,
+    },
+    content: html`
+      <div class="card mb-4">
+        <div class="card-header bg-primary text-white d-flex align-items-center">
+          <h1>Assessments</h1>
+        </div>
+
+        <div class="table-responsive">
+          <table class="table table-sm table-hover">
+            <thead>
+              <tr>
+                <th style="width: 1%"><span class="visually-hidden">Label</span></th>
+                <th><span class="visually-hidden">Title</span></th>
+                <th>AID</th>
+              </tr>
+            </thead>
+            <tbody>
+              ${rows.map(
+                (row) => html`
+                  ${row.start_new_assessment_group
+                    ? html`
+                        <tr>
+                          <th colspan="3" scope="row">
+                            ${assessmentsGroupBy === 'Set'
+                              ? AssessmentSetHeading({ assessment_set: row.assessment_set })
+                              : AssessmentModuleHeading({
+                                  assessment_module: row.assessment_module,
+                                })}
+                          </th>
+                        </tr>
+                      `
+                    : ''}
+                  <tr id="row-${row.id}">
+                    <td class="align-middle" style="width: 1%">
+                      <span class="badge color-${row.assessment_set.color}"> ${row.label} </span>
+                    </td>
+                    <td class="align-middle">
+                      <a
+                        href="/pl/public/course_instance/${resLocals.course_instance_id}/assessment/${row.id}/questions"
```

Comment:
Was this tested? `course_instance_id` isn't a property on `resLocals`, so this produces a URL like `/pl/public/course_instance//assessment/5184/questions` which is invalid.

As with all new pages, I'd advise against reaching into `res.locals` if you can avoid it. I know it may still be needed for the navbar, but you can still pass a typed `course_instance` object separately to this `PublicAssessments` function.

Reply:
oh no! I did test this working at one point but I must have broken it while moving stuff around (all the more reason to not use resLocals lol). 

fixed in dd2e51b894

Reply:
Non-blocking, but I suggested a whole `course_instance` object for a reason: a random string ID is not inherently trusted, especially when it's coming from `req.params` (though this isn't as big of a deal since this is a public page). On the other hand, having a fully-typed row object means that the DB has confirmed that thing X exists, and most of our selection functions will validate that they belong to an enclosing course or whatever.

Reply:
I added proper validation for this